README.md
Maximum Likelihood Estimation of a Truncated Normal Distribution with
Censored Data
================

<!-- badges: start -->

[![CRAN
version](http://www.r-pkg.org/badges/version/tcensReg)](http://www.r-pkg.org/pkg/tcensReg)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/grand-total/tcensReg)](http://www.r-pkg.org/pkg/tcensReg)
[![cran
checks](https://cranchecks.info/badges/worst/tcensReg)](https://cranchecks.info/pkgs/tcensReg)
<!-- badges: end -->

The goal of this package is to estimate parameters from a linear model
when the data comes from a truncated normal distribution with censoring.
Maximum likelihood values are returned. There are multiple method
available for optimization with the default set as conjugate gradient.
This package is also able to return maximum likelihood estimates for
truncated only or censored only data similar to `truncreg` and `censReg`
packages.

For details on the methods, simulation performance, and an application
to visual quality testing from clinical trials for intraocular lenses
see the corresponding manuscript with open source access available at
[doi:10.1186/s12874-020-01032-9](https://doi.org/10.1186/s12874-020-01032-9).

# Installation

You can install `tcensReg` from CRAN for the stable release or install
the GitHub version for the active development version:

``` r
#stable CRAN version
install.packages("tcensReg")

# or ----

#active devel. GitHub version
install.packages("devtools")
devtools::install_github("williazo/tcensReg")
```

# Example 1: Single Population

Some common examples where this type of problem may arise is when there
is a natural truncation imposed by the structure of the data. For
instance several applications have an implied zero truncation such as
product lifetimes, age, or detection thresholds. To show how to
implement the functions within the package, I will demonstrate a simple
simulation example.

Assume that we have observations from an underlying truncated normal
distribution. In our case we will assume a zero-truncated model by
setting a=0. We generate this truncated normal data below and refer to
it as `y_star`.

``` r
library(tcensReg)
mu <- 0.5
sigma <- 0.5
a <- 0

set.seed(032420)
#generate random values from the truncated normal distribution using tcensReg function
y_star <- rtnorm(n=1000, mu=mu, sd=sigma, a=a)
#note that the lowerbound will always be non-negative
round(range(y_star), 3)
```

    ## [1] 0.004 2.217

Next, we can imagine a scenario where we have an imprecise measurement
of `y_star` leading to censoring. In our case we assume that values
below a limit of detection, `nu`, are censored. This creates a random
variable `y`.

In the example below we set our limit of detection as `nu`=0.25.

``` r
nu <- 0.25
y <- ifelse(y_star <= nu, nu, y_star)
#calculating the number of censored observations
sum(y == nu)/length(y) 
```

    ## [1] 0.199

``` r
#collecting the uncensored and censored data together
dt <- data.frame(y_star, y) 
```

We can observe the histogram and density plot for the uncensored data,
which shows the zero-truncation.
![](README_files/figure-gfm/univariate_uncens_hist-1.png)<!-- -->

We can then compare this to the censored observations below
![](README_files/figure-gfm/univariate_cens_hist-1.png)<!-- -->

We can then estimate the mean, `mu`, and standard deviation, `sigma`,
using `y` with the `tcensReg` package as shown below.

``` r
#loading the package
library(tcensReg)  
t_mod <- tcensReg(y ~ 1, data=dt, a=0, v=0.25)
summary(t_mod)
```

    ## 
    ## Call:
    ## tcensReg(formula = y ~ 1, a = 0, v = 0.25, data = dt)
    ## 
    ## Assumed Distribution:
    ## Truncated Normal with Censoring
    ## 
    ## Count of Observations:
    ##    Total Observations Censored Observations 
    ##                  1000                   199 
    ## 
    ## 
    ## Coefficients:
    ##             Estimate Std. Error t value
    ## (Intercept)   0.4562     0.0359 12.7024
    ## sigma         0.5405     0.0142 38.0508
    ## 
    ## Log Likelihood: -710.7033
    ## Information Criterion: AIC=1425.4066 BIC=1435.2221
    ## Optimization Method: CG
    ## Psuedo R2: 0 method - mckelvey_zavoina

By default the coefficients are returned along with log likelihood and
other fit criterion statistics. Note that the Pseudo R2 in the case of
an intercept only model is exactly equal to zero.

``` r
names(t_mod)
```

    ##  [1] "theta"             "convergence"       "initial_ll"       
    ##  [4] "final_ll"          "var_cov"           "method"           
    ##  [7] "info_criteria"     "model_matrix"      "call"             
    ## [10] "n_count"           "latent_assumption"

Note that the this object contains parameter estimates theta,
convergence criterion code, initial/final log-likelihood values,
variance-covariance matrix, method of optimization, information
criterion, design matrix used from the model, formula call, count of
total/censored observations, and latent distributional assumption.

Comparing the values to the truth we see that the estimates are
unbiased.

``` r
#tcensReg model
output <- tcensReg(y ~ 1, data=dt, a=a, v=nu)
#extracting the point estimates
tcensReg_est <- coef(output) #this returns sigma rather than log sigma

#OLS model
lm_output <- lm(y ~ 1, data=dt) 
lm_est <- c(coef(lm_output), summary(lm_output)$sigma)
#censored only model, i.e., Tobit model
cens_output <- tcensReg(y ~ 1, data=dt, v=nu) 
```

    ## Warning: `a` is not specified indicating no truncation

``` r
cens_est <- coef(cens_output)

results_df <- data.frame(rbind(c(mu, sigma),
                               t(tcensReg_est),
                               lm_est,
                               t(cens_est)))
names(results_df) <- c("mu", "sigma")
row.names(results_df) <- c("Truth", "tcensReg", "Normal MLE", "Tobit")
results_df$mu_bias <- results_df$mu - mu
results_df$sigma_bias <- results_df$sigma - sigma

knitr::kable(results_df, format="markdown", digits=4)
```

|            |     mu |  sigma | mu\_bias | sigma\_bias |
| :--------- | -----: | -----: | -------: | ----------: |
| Truth      | 0.5000 | 0.5000 |   0.0000 |      0.0000 |
| tcensReg   | 0.4562 | 0.5405 | \-0.0438 |      0.0405 |
| Normal MLE | 0.6685 | 0.3841 |   0.1685 |    \-0.1159 |
| Tobit      | 0.6166 | 0.4595 |   0.1166 |    \-0.0405 |

Other methods result in significant bias for both `mu` and `sigma`.

# Example 2: Two Population Model with Separate Variances

As an extension for the single population model above, we can imagine a
two independent truncated normal random variables that have common
censoring and truncation values but different standard deviations.

We can simulate the underlying truncated normal distributions `Y1_star`
and `Y2_star` similar to \(Y\) above except now we allow them to have
separate mean and variances.

For this example we let `mu_1`=0.5, `mu_2`=1, `sigma_1`=0.25,
`sigma_2`=2, and `a`=0.

``` r
mu_1 <- 0.5
mu_2 <- 1
sigma_1 <- 0.25
sigma_2 <- 2
a <- 0

set.seed(032420)
y_1_star <- rtnorm(1000, mu = mu_1, sd = sigma_1, a = a)
y_2_star <- rtnorm(1000, mu = mu_2, sd = sigma_2, a = a)
df <- data.frame(y_star = c(y_1_star, y_2_star), 
                 group = c(rep("Population 1", length(y_1_star)),
                           rep("Population 2", length(y_2_star))))
```

Plotting each of these uncensored population densities, we can see the
difference in shape based on the underlying parameter selection.

![](README_files/figure-gfm/two_pop_graph-1.png)<!-- -->

Then censoring each observation at `nu`, we are left with `Y1` and `Y2`.
Again, we let `nu`=0.25.

``` r
nu <- 0.25
df$y <- ifelse(df$y_star<=nu, nu, df$y_star)
```

We then can fit our model with separate variances for each group using
the command `tcensReg_sepvar` as shown below.

``` r
mod_result <- tcensReg_sepvar(y ~ group, a=a, v=nu, group_var="group", method="maxLik", data=df)
mod_result
```

    ## $theta
    ##       (Intercept) groupPopulation 2        log_sigma1        log_sigma2 
    ##         0.4933551         0.6659116        -1.3281990         0.6546893 
    ## 
    ## $convergence
    ## [1] 2
    ## 
    ## $initial_ll
    ## [1] -2013.046
    ## 
    ## $final_ll
    ## [1] -1933.59
    ## 
    ## $var_cov
    ##                     (Intercept) groupPopulation 2    log_sigma1    log_sigma2
    ## (Intercept)        9.384525e-05     -9.384525e-05 -9.865163e-05 -1.179984e-14
    ## groupPopulation 2 -9.384525e-05      2.475291e-02  9.865163e-05 -6.214006e-03
    ## log_sigma1        -9.865163e-05      9.865163e-05  8.721167e-04  1.240418e-14
    ## log_sigma2        -1.179984e-14     -6.214006e-03  1.240418e-14  2.211993e-03
    ## 
    ## $method
    ## [1] "maxLik"

``` r
sepvar_est <- mod_result$theta
mu_1_est <- sepvar_est[1]
mu_2_est <- sum(sepvar_est[1:2])
sigma_1_est <- exp(sepvar_est[3])
sigma_2_est <- exp(sepvar_est[4])

results_df <- data.frame(rbind(c(mu_1, mu_2, sigma_1, sigma_2),
                               c(mu_1_est, mu_2_est, sigma_1_est, sigma_2_est)))
names(results_df) <- c("mu_1", "mu_2", "sigma_1", "sigma_2")
row.names(results_df) <- c("Truth", "tcensReg")
results_df$mu1_pct_bias <- paste0(round(((results_df$mu_1 - mu_1)/mu_1)*100, 2), "%")
results_df$mu2_pct_bias <- paste0(round(((results_df$mu_2 - mu_2)/mu_2)*100, 2), "%")
results_df$sigma1_pct_bias <- paste0(round(((results_df$sigma_1 - sigma_1)/sigma_1)*100, 2), "%")
results_df$sigma2_pct_bias <- paste0(round(((results_df$sigma_2 - sigma_2)/sigma_2)*100, 2), "%")

knitr::kable(results_df, format="markdown", digits=4)
```

|          |  mu\_1 |  mu\_2 | sigma\_1 | sigma\_2 | mu1\_pct\_bias | mu2\_pct\_bias | sigma1\_pct\_bias | sigma2\_pct\_bias |
| :------- | -----: | -----: | -------: | -------: | :------------- | :------------- | :---------------- | :---------------- |
| Truth    | 0.5000 | 1.0000 |    0.250 |   2.0000 | 0%             | 0%             | 0%                | 0%                |
| tcensReg | 0.4934 | 1.1593 |    0.265 |   1.9245 | \-1.33%        | 15.93%         | 5.98%             | \-3.77%           |

# Performance Comparison: Censored-Only and Truncated-Only

Note also that the `tcensReg` can also estimate parameters in the
censored-only or truncated-only cases. We show below that by using
analytic values in the tcensReg implementation that our method is faster
then the alternative estimation procedures while providing better
variance estimates. With a small set of covariates and `p<<n` we can use
the Newton-Raphson method of optimization, which is computationally fast
with few covariates.

``` r
library(microbenchmark)
#testing the censored-only regression
library(censReg)
cens <- microbenchmark(tcensReg_method = tcensReg(y ~ 1, data=dt, v=nu, method="Newton"),
               censReg_method = censReg(y ~ 1, left=nu, data=dt))
knitr::kable(summary(cens), format="markdown", digits=4)
```

| expr             |     min |      lq |    mean |  median |      uq |      max | neval | cld |
| :--------------- | ------: | ------: | ------: | ------: | ------: | -------: | ----: | :-- |
| tcensReg\_method |  4.8881 |  5.0943 |  5.9877 |  5.2487 |  5.7981 |  14.1300 |   100 | a   |
| censReg\_method  | 13.7873 | 14.5127 | 19.2472 | 18.2919 | 21.4700 | 103.8775 |   100 | b   |

``` r
#point estimates are equivalent
tcensReg_est <- as.numeric(tcensReg(y ~ 1, data=dt, v=nu, method="Newton")$theta)
censReg_est <- as.numeric(coef(censReg(y ~ 1, left=nu, data=dt)))
all.equal(tcensReg_est, censReg_est)
```

    ## [1] TRUE

``` r
#testing the truncated-only regression
library(truncreg)
trunc <- microbenchmark(
  tcensReg_method = tcensReg(y_star ~ 1, data=dt, a=a, method="Newton"),
  truncreg_method = truncreg(y_star ~ 1, point=a, data=dt))
knitr::kable(summary(trunc), format="markdown", digits=4)
```

| expr             |     min |      lq |    mean |  median |      uq |     max | neval | cld |
| :--------------- | ------: | ------: | ------: | ------: | ------: | ------: | ----: | :-- |
| tcensReg\_method |  9.4831 |  9.6810 | 11.3402 |  9.9553 | 12.1703 | 19.8827 |   100 | a   |
| truncreg\_method | 16.2908 | 16.7846 | 19.4056 | 17.5601 | 21.6932 | 35.6788 |   100 | b   |

``` r
tcensReg_est <- as.numeric(tcensReg(y_star ~ 1, data=dt, a=a, method="Newton")$theta)
#note truncreg returns sigma not log_sigma so we need to exponentiate our value
tcensReg_est[2] <- exp(tcensReg_est[2])
truncreg_est <- as.numeric(coef(truncreg(y_star ~ 1, point=a, data=dt)))
all.equal(tcensReg_est, truncreg_est)
```

    ## [1] "Mean relative difference: 0.0003643991"

In the comparisons above we are using an intercept only model, but in
general we expect that interest lies in understanding how a set of
covariates effect the mean response. So to test the sensitivity and
speed as the number of covariates approaches `n` we can generate
independent random variables `X` and fit the regression model of `Y` or
`Y_star`.

We can compare the censored-only and truncated-only performance with 100
predictors, i.e. `p`=20. To illustrate some of the other available
optimization methods we will set method to BFGS, which is a quasi-Newton
optimization method.

``` r
#number of predictors
p <- 20
X <- NULL
for(i in seq_len(p)){
    X_i <- rnorm(n = length(y))
    X <- cbind(X, X_i)
}
colnames(X) <- paste0("var_", seq_len(p))
dt <- data.frame(y, X)

#testing the censored-only regression with 100 covariates
cens <- microbenchmark(tcensReg_method = tcensReg(y ~ ., data=dt, v=nu, method="BFGS"),
               censReg_method = censReg(y ~ ., left=nu, data=dt))
knitr::kable(summary(cens), format="markdown", digits=4)
```

| expr             |      min |       lq |     mean |   median |       uq |      max | neval | cld |
| :--------------- | -------: | -------: | -------: | -------: | -------: | -------: | ----: | :-- |
| tcensReg\_method | 218.2029 | 223.9493 | 234.8069 | 226.8067 | 230.3986 | 366.2588 |   100 | a   |
| censReg\_method  | 345.9823 | 362.4582 | 386.9384 | 368.1871 | 373.6284 | 530.3745 |   100 | b   |

``` r
#point estimates are equivalent
tcensReg_est <- as.numeric(tcensReg(y ~ ., data=dt, v=nu, method="BFGS")$theta)
censReg_est <- as.numeric(coef(censReg(y ~ ., left=nu, data=dt)))
all.equal(tcensReg_est, censReg_est)
```

    ## [1] "Mean relative difference: 8.089508e-05"

``` r
#testing the truncated-only regression with 100 covariates
trunc <- microbenchmark(tcensReg_method = tcensReg(y_star ~ ., data=dt, a=a, method="BFGS"),
                        truncreg_method = truncreg(y_star ~ ., point=a, data=dt))
knitr::kable(summary(trunc), format="markdown", digits=4)
```

| expr             |      min |       lq |     mean |   median |       uq |      max | neval | cld |
| :--------------- | -------: | -------: | -------: | -------: | -------: | -------: | ----: | :-- |
| tcensReg\_method | 202.3384 | 207.2465 | 215.1095 | 210.5763 | 213.9136 | 363.9193 |   100 | a   |
| truncreg\_method | 461.7276 | 470.2840 | 490.9210 | 475.7646 | 480.7120 | 637.6223 |   100 | b   |

``` r
tcensReg_est <- as.numeric(tcensReg(y_star ~ ., data=dt, a=a, method="BFGS")$theta)
#note truncreg returns sigma not log_sigma so we need to exponentiate the last parameter value
tcensReg_est[length(tcensReg_est)] <- exp(tcensReg_est[length(tcensReg_est)])
truncreg_est <- as.numeric(coef(truncreg(y_star ~ ., point=a, data=dt)))
all.equal(tcensReg_est, truncreg_est)
```

    ## [1] "Mean relative difference: 2.993568e-05"


NEWS
version 0.1.7 (2020-06-29)
	+ update citation for publication of article in BMC Medical Research Methodology

version 0.1.6 (2020-03-27)
	+ create S3 class "tcensReg" for output from tcensReg() function
		+ available S3 methods include: print, coef, predict, summary, and vcov
	+ added pseudo_r2() function to calculate goodness of fit measures for censored observations
	+ updates to documentation for functions for added clarity
		+ also added references and equations
	+ removed internal functions from exported documentation such
		+ internal functions include those used to evaluate likelihood, gradient, hessian, or direct optimization procedure
	+ added simulation code for simulation study 4 for behavior as sample size increases

version 0.1.5 (2020-01-03)
	+ updates to rtnorm() function to correct errors with vector values mean and scalar
	+ adding in inst/script function into CRAN version after updating future_lapply() and package installation
	+ updates to vignette for installation guide for CRAN

version 0.1.4: (2019-12-18)
	+ package accepted into CRAN repository


R/tcensReg.R
#'  Regression Method for Truncated Normal Distribution with Censored Data
#'
#'  This function is used to find estimates from a linear equation
#'  assuming that the underlying distribution is truncated normal
#'  and the data has subsequently been censored data.
#'
#' @param formula Object of class \code{formula} which symbolically describes the model to be fit
#' @param a Numeric scalar indicating the truncation value. Initial value is -Inf indicating no truncation
#' @param v Numeric scalar indicating the censoring value. Initially set to NULL indicating no censoring
#' @param data Data.frame that contains the outcome and corresponding covariates. If none is provided then assumes objects are in user's environment.
#' @param method Character value indicating which optimization routine to perform.
#' Choices include \code{Newton}, \code{BFGS}, and \code{CG}. See details for explanation on each method.
#' @param ... Additional arguments from such as \code{max_iter}, \code{step_max}, or \code{epsilon}.
#' See details for how to define these additional arguments.
#'
#' @details
#'  This estimation procedure returns maximum likelihood estimates under the presence
#'  of \emph{left} truncation and/or \emph{left} censoring. It builds upon currently available methods
#'  for limited dependent variables by relaxing the assumption of a latent normal distribution
#'  and instead allows the this underlying distribution to potentially be a
#'  latent \strong{truncated} normal distribution.
#'
#'  To indicate left censoring the user should specify the parameter \code{v}, and
#'  to indicate left truncation specify the parameter \code{a}. If specifying
#'  both left censoring and left truncation note that there is an implicit restriction that
#'  \eqn{a<\nu}.
#'
#'  Below is a brief description of the types of distributions that can be fit
#'  along with the assumed data generating process for the observed outcome, \eqn{Y}.
#'
#'  \subsection{Latent Distributions}{
#'
#'  The tcensReg function allows user to specify one of four combinations of
#'  distributional assumptions with or without censoring. These are listed below
#'  along with the necessary arguments needed to fit this model.
#'    \subsection{Truncated Normal with Censoring}{
#'
#'    This is the main model that this package is designed to fit and introduced
#'    in \insertCite{williams2019modeling}{tcensReg}.It assumes
#'    \deqn{Y_{i}^{*}\sim TN(\mu, \sigma^{2}, a)}
#'    where TN indicates a left truncated normal random variable, truncated at
#'    the value \eqn{a}.
#'
#'    This underlying truncated normal random variable is then \emph{left} censored at the value
#'    \eqn{\nu} to create the censored observations \eqn{Y} such that
#'    \deqn{Y_{i}=\nu 1_{\{Y_{i}^{*}\le\nu\}}
#'    + Y_{i}^{*}
#'    1_{\{Y_{i}^{*}>\nu}\}}
#'
#'    Required Arguments:
#'    \itemize{
#'      \item \code{a}: left truncation value
#'      \item \code{v}: left censoring value
#'      }
#'    }
#'    \subsection{Normal with Censoring}{
#'
#'    This model is commonly referred to as the Tobit model
#'    \insertCite{tobin1958estimation}{tcensReg}. This model assumes that the
#'    data is generated from a latent normal random variable \eqn{Y_{i}^{*}}, i.e.,
#'    \deqn{Y_{i}^{*}\sim N(\mu, \sigma^{2})}
#'
#'    This underlying normal random variable is then \emph{left} censored at the value
#'    \eqn{\nu} to create the censored observations \eqn{Y} such that
#'    \deqn{Y_{i}=\nu 1_{\{Y_{i}^{*}\le\nu\}}
#'    + Y_{i}^{*}
#'    1_{\{Y_{i}^{*}>\nu}\}}
#'
#'    Required Arguments:
#'    \itemize{
#'     \item \code{v}: left censoring value
#'    }
#'
#'    This procedure can also be fit using the \code{\link[censReg]{censReg}} package by
#'    \insertCite{henningsen2010estimating;textual}{tcensReg}.
#'    }
#'    \subsection{Truncated Normal}{
#'
#'      This model assumes that there is no censored observations, but that the
#'      data are left truncated as originally described by
#'      \insertCite{hald1949maximum;textual}{tcensReg}.
#'
#'      Therefore, we assume that the observed values follow
#'      \deqn{Y_{i}^\sim TN(\mu, \sigma^{2}, a)}
#'      where TN indicates a left truncated normal random variable, truncated at
#'      the value \eqn{a}.
#'
#'      Required Arguments:
#'      \itemize{
#'       \item \code{a}: left truncation value
#'      }
#'
#'      This procedure can also be fit using the \code{\link[truncreg]{truncreg}} package by
#'      \insertCite{croissant2018truncreg;textual}{tcensReg}.
#'     }
#'     \subsection{Normal}{
#'
#'     This model assumes that there is no left censoring and no left truncation.
#'
#'     Maximum likelihood estimates are returned based on the assumption that the
#'     random variable follows the distribution
#'     \deqn{Y_{i}^\sim N(\mu, \sigma^{2})}
#'
#'     Required Arguments: None
#'
#'     This procedure can also be fit using the command \code{\link{lm}} in base R.
#'     }
#'  }
#'
#'  \subsection{Optimization Methods}{
#'
#'  Currently available optimization routines include conjugate gradient
#'  (\code{CG}), Newton-Raphson (\code{Newton}), and BFGS (\code{BFGS}).
#'  The default method is set as the conjugate gradient. Both the of the
#'  conjugate gradient and BFGS methods are implemented via the
#'  general-purpose optimization routine \code{\link{optim}}. These two methods
#'  use only the respective likelihood and gradient functions.
#'  The Newton-Raphson method uses the likelihood, gradient, and Hessian
#'  functions along with line search to achieve the maximum likelihood.
#'  }
#'
#'  \subsection{Additional Arguments}{
#'
#'  There are additional arguments that the user may provide for controlling the
#'  optimization routine.
#'  \itemize{
#'    \item{max_iter: Maximum number of iterations for optimization routine. Default is 100}
#'    \item{step_max: Maximum number of steps when performing line search. Default is 10}
#'    \item{epsilon: Numeric value used to define algorithm stops, i.e., when evaluated gradient is less than epsilon. Default is 0.001}
#'    \item{tol_val: Tolerance value used to stop the algorithm if the (n+1) and (n) log likelihood is within the tolerance limit.}
#'    \item{theta_init: Numeric vector specifying the initial values to use for the estimated parameters \eqn{\beta} and \eqn{\log(\sigma)}}
#'  }
#'  }
#'
#' @importFrom stats model.frame model.matrix
#'
#' @examples
#' #truncated normal underlying data
#' y_star <- rtnorm(n = 1000, mu = 0.5, sd = 1, a = 0)
#'
#' #apply censoring
#' y <- ifelse(y_star <= 0.25, 0.25, y_star)
#'
#' #find MLE estimates
#' trunc_cens_mod <- tcensReg(y ~ 1, v = 0.25, a = 0)
#' summary(trunc_cens_mod)
#'
#'
#' @return Returns a list of final estimate of theta, total number of iterations performed, initial log-likelihood,
#' final log-likelihood, estimated variance covariance matrix, information criterion, model design matrix,
#' call, list of total observations and censored observations, and latent distributional assumption.
#'
#' @export
#' @references
#'     \insertAllCited{}
tcensReg <- function(formula, a = -Inf, v = NULL, data = sys.frame(sys.parent()),
                     method = c("CG", "Newton", "BFGS"), ...){
    #checks for proper specification of formula
    method <- match.arg(method)
    if(class(formula) != "formula"){
        stop("`formula` must be a formula", call. = FALSE)
        } else if (length(formula) != 3){
            stop("`formula` must be 2-sided", call. = FALSE)
        }

    #outcome vector
    y <- model.frame(formula, data)[, 1]
    #design matrix
    X <- model.matrix(formula, data)

    #observed censoring
    n_total <- length(y)
    n_cens <- sum(y==v)

    #checking for proper specification of a and v
    if(!is.null(v) & (!is.numeric(a) | !is.numeric(v))){
        stop("`a` and `v` must both be numeric", call. = FALSE)
        } else if(!is.null(v) & (length(a) !=1 | length(v) != 1)){
            stop("`a`, and `v` must both be scalars", call. = FALSE)
            }
    # checking model specification of truncation and censoring params
    if(is.null(v) & a == -Inf){
        warning("`v` and `a` are not specified indicating no censoring and no truncation", call. = FALSE)
        } else if(is.null(v) & a != -Inf){
            if(any(y < a)){
                stop("observed values below specified truncation `a`", call.=FALSE)
            }
            warning("`v`is not specified indicating no censoring", call. = FALSE)
            } else if(!is.null(v) & a == -Inf){
                len_cens <- sum(y == v, na.rm=TRUE)
                if(len_cens == 0){
                    stop("censoring indicated but no observed censored values",
                         call.=FALSE)
                    }
                warning("`a` is not specified indicating no truncation", call. = FALSE)
                } else {
                    if(v < a) stop("censoring specified below truncation", call.=FALSE)
                    if(any(y < a)){
                        stop("observed values below specified truncation `a`", call.=FALSE)
                    }
                    len_cens <- sum(y == v, na.rm = TRUE)
                    if(len_cens == 0){
                        stop("censoring indicated but no observed censored values",
                             call.=FALSE)
                    }
                }

    #reading in the newton raphson for the truncated censored normal
    if(method == "Newton"){
        results <- tcensReg_newton(y, X, a, v, ...)
    } else if(method %in%c("BFGS", "CG")){
        results <- tcensReg_optim(y, X, a, v, method, ...)
    }

    #adding in information criterion into the results
    aic <- (2 * length(results$theta)) - (2 * results$final_ll)
    bic <- log(length(y)) * length(results$theta) - 2 * results$final_ll
    info_criteria <- c(AIC=aic, BIC=bic)
    results[["info_criteria"]] <- info_criteria

    #additional information to save in order to use S3 methods
    results$model_matrix <- X
    results$call <- match.call()
    results$n_count <- list(n=n_total, n_cens=n_cens)
    latent_assumption <- ifelse(is.null(v) & a == -Inf, "Normal",
                                ifelse(is.null(v) & a != -Inf, "Truncated Normal",
                                       ifelse(!is.null(v) & a == -Inf, "Normal with Censoring",
                                              ifelse(!is.null(v) & a != -Inf, "Truncated Normal with Censoring"))))
    results$latent_assumption <- latent_assumption
    #saving object as S3
    class(results) <- "tcensReg"
    return(results)
}
